### PR TITLE
KK-401 | Free spot notifications implementation and tests

### PR DESCRIPTION
**Pull request by tuomas777**
Added implementation for free spot notifications. The implementation should comply with [these specs](https://helsinkisolutionoffice.atlassian.net/browse/KK-245).

The PR also includes tests for subscriptions, and one new API error for a situation when there are still free spots when one tries to subscribe.

### CODE REVIEW

**Review by quyenlq — APPROVED** [review]
Nice work! Left small comments but overall it's beautiful 👏

File: subscriptions/factories.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import factory
+from subscriptions.models import FreeSpotNotificationSubscription
+
+from children.factories import ChildFactory
+from events.factories import EnrolmentFactory, OccurrenceFactory
+
+
+def create_occurrence():
```
[the remark this answers is not in the document]
Comment (tuomas777) [reply subscriptions/factories.py]: Yes that is exactly what happened, good spot! 🙂

File: subscriptions/models.py [hunk 2]
```diff
@@ -12,6 +16,20 @@ def user_can_view(self, user):
...
         ).distinct()
 
+    def send_notification(self):
+        for subscription in self.select_related("occurrence__event"):
+            child = subscription.child
+            occurrence = subscription.occurrence
+
+            subscription.delete()
+
+            send_event_notifications_to_guardians(
```
[the remark this answers is not in the document]
Comment (tuomas777) [reply subscriptions/models.py:26]: Good point! However, I think this is a bit tricky with no perfect solution. In a possible error situation, there are two possible things of which we can ensure only one:

1) make sure the subscription is deleted even if that means no notification will be sent
2) make sure the notification is sent even if that means the subscription will not be deleted

I kind of think that failing to send a single notification (1) is less bad than risking ending up spamming multiple notifications of a subscription in "erroneous" state (2). Dunno.

[the remark this answers is not in the document]
Comment (quyenlq) [reply subscriptions/models.py:26]: Makes sense. It'd be the best if `send_event_..` return a flag to indicate if the sending process successfully, then we can delete the subscription depends on that returned value, but it's a bit overkill to handle this corner case. For now, I think this looks fine :)

**Pull request merged**